tests/devices/qubit/test_simulate.py

```diff
@@ -0,0 +1,583 @@
+from pennylane.devices.qubit import simulate
+
+ml_frameworks_list = [
+    "numpy",
+    pytest.param("autograd", marks=pytest.mark.autograd),
+    pytest.param("jax", marks=pytest.mark.jax),
+    pytest.param("torch", marks=pytest.mark.torch),
+    pytest.param("tensorflow", marks=pytest.mark.tf),
+]
+
+
+class TestCurrentlyUnsupportedCases:
+    def test_hamiltonian_observable(self):
+        """Test that measuring hamiltonians gives a NotImplementedError."""
+
+        H = qml.Hamiltonian([2.0], [qml.PauliX(0)])
+        qs = qml.tape.QuantumScript(measurements=[qml.expval(H)])
+        with pytest.raises(NotImplementedError):
+            simulate(qs)
+
+    def test_sample_based_observable(self):
+        """Test sample-only measurements raise a notimplementedError."""
+
+        qs = qml.tape.QuantumScript(measurements=[qml.sample(wires=0)])
+        with pytest.raises(NotImplementedError):
+            simulate(qs)
+
+
+def test_custom_operation():
+    """Test execution works with a manually defined operator if it has a matrix."""
+
+    # pylint: disable=too-few-public-methods
+    class MyOperator(qml.operation.Operator):
+        num_wires = 1
+
+        @staticmethod
+        def compute_matrix():
+            return qml.PauliX.compute_matrix()
+
+    qs = qml.tape.QuantumScript([MyOperator(0)], [qml.expval(qml.PauliZ(0))])
+
+    result = simulate(qs)
+    assert qml.math.allclose(result[0], -1.0)
+
+
+# pylint: disable=too-few-public-methods
+class TestStatePrep:
+    """Tests integration with various state prep methods."""
+
+    def test_basis_state(self):
+        """Test that the basis state method operator the quantum script into the right state."""
+        qs = qml.tape.QuantumScript(
+            measurements=[qml.probs(wires=(0, 1, 2))], prep=[qml.BasisState([0, 1], wires=(0, 1))]
+        )
+        probs = simulate(qs)[0]
+        expected = np.zeros(8)
+        expected[2] = 1.0
+        assert qml.math.allclose(probs, expected)
+
+
+class TestBasicCircuit:
+    """Tests a basic circuit with one rx gate and two simple expectation values."""
+
+    def test_basic_circuit_numpy(self):
+        """Test execution with a basic circuit."""
+        phi = np.array(0.397)
+        qs = qml.tape.QuantumScript(
+            [qml.RX(phi, wires=0)], [qml.expval(qml.PauliY(0)), qml.expval(qml.PauliZ(0))]
+        )
+        result = simulate(qs)
+
+        assert isinstance(result, tuple)
+        assert len(result) == 2
+
+        assert np.allclose(result[0], -np.sin(phi))
+        assert np.allclose(result[1], np.cos(phi))
+
+    @pytest.mark.autograd
+    def test_autograd_results_and_backprop(self):
+        """Tests execution and gradients with autograd"""
+        phi = qml.numpy.array(-0.52)
+
+        def f(x):
+            qs = qml.tape.QuantumScript(
+                [qml.RX(x, wires=0)], [qml.expval(qml.PauliY(0)), qml.expval(qml.PauliZ(0))]
+            )
+            return qml.numpy.array(simulate(qs))
```

Comment:
Is the `qml.numpy.array` required? The output of `simulate` should be an array, right?

Reply:
Autograd can't differentiate a tuple of arrays.